internal/gribi/gribi.go

```diff
@@ -0,0 +1,203 @@
+type GRIBIHandler struct {
+	dut            *ondatra.DUTDevice
+	fluentC        *fluent.GRIBIClient
+	fibACK         bool
+	persistence    bool
+	redundancyMode fluent.RedundancyMode
+	lastElectionID *spb.Uint128
+}
+
+var mu sync.Mutex
+
+// NewGRIBIFluent establishes a gribi client session with the dut.
+// persistence and fibACK specify how the session to be established.
+// The client connection handles (fluent client) and session parameters are cached.
+// The client is not the leader by default and for that function BecomeLeader needs to be called.
+func NewGRIBIFluent(t testing.TB, dut *ondatra.DUTDevice, persistence, fibACK bool) *GRIBIHandler {
+	fluentClient := fluent.NewClient()
+	gribiC := dut.RawAPIs().GRIBI().Default(t)
+	fluentClient.Connection().WithStub(gribiC)
+	g := &GRIBIHandler{dut: dut,
+		fluentC:     fluentClient,
+		fibACK:      fibACK,
+		persistence: persistence,
+		// we assume all of test cases use ElectedPrimaryClient
+		redundancyMode: fluent.ElectedPrimaryClient,
+		lastElectionID: &spb.Uint128{Low: 1,
+			High: 0,
+		},
+	}
+	if persistence {
+		fluentClient.Connection().WithInitialElectionID(g.lastElectionID.Low, g.lastElectionID.High).WithRedundancyMode(fluent.ElectedPrimaryClient).WithPersistence()
+	} else {
+		fluentClient.Connection().WithInitialElectionID(g.lastElectionID.Low, g.lastElectionID.High).WithRedundancyMode(fluent.ElectedPrimaryClient)
+	}
+	if fibACK {
+		fluentClient.Connection().WithFIBACK()
+	}
+	fluentClient.Start(context.Background(), t)
+	fluentClient.StartSending(context.Background(), t)
+	err := g.AwaitTimeout(context.Background(), t, timeout)
+	if err != nil {
+		t.Fatalf("can not establish the session: %v", err)
+	}
+	return g
+}
+
+// Fluent resturns the fluent client that can be used to directly call the gribi fluent apis
+func (g *GRIBIHandler) Fluent(t testing.TB) *fluent.GRIBIClient {
+	return g.fluentC
+}
+
+// Close function closes the gribi session with the dut by stopping the fluent client.
+func (g *GRIBIHandler) Close(t testing.TB) {
+	t.Helper()
+	t.Logf("closing GRIBI connection for dut: %s", g.dut.Name())
+	g.fluentC.Stop(t)
+}
+
+// AwaitTimeout calls a fluent client Await by adding a timeout to the context.
+func (g *GRIBIHandler) AwaitTimeout(ctx context.Context, t testing.TB, timeout time.Duration) error {
+	subctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+	return g.fluentC.Await(subctx, t)
+}
+
+// LearnElectionID learns the current server election id of the dut by sending a dummy modify request. The function caches the election id.
+func (g *GRIBIHandler) LearnElectionID(t testing.TB) (low, high uint64) {
+	t.Helper()
+	t.Logf("learn GRIBI Election ID from dut: %s", g.dut.Name())
+	g.fluentC.Modify().UpdateElectionID(t, g.lastElectionID.Low, g.lastElectionID.High)
+	err := g.AwaitTimeout(context.Background(), t, timeout)
+	if err != nil {
+		t.Fatalf("learnElectionID Error: %v", err)
+	}
+	results := g.fluentC.Results(t)
+	mu.Lock()
```

Comment:
Thanks for pointing this out. I was trying to use lock because of election id.  However, I noticed the gribigo client already saves the last known id, so I removed election id from the struct and the locking as well.